examples/gno.land/p/demo/timelock/timelock.gno

```diff
@@ -0,0 +1,150 @@
+//  id := seqid.ID()
+//  timelockUtil.Schedule(id, 60, 30)
+//
+//  Check if an operation is pending.
+//  isPending := timelockUtil.IsOperationPending(id)
+//
+//  Execute the operation when it is ready.
+//  if timelockUtil.IsOperationReady(id) {
+//      timelockUtil.Execute(id)
+//  }
+//
+//  Update the minimum delay for future operations.
+//  var minDelay int64 = 30
+//  timelockUtil.UpdateDelay(&minDelay, 45)
+
+package timelock // import "gno.land/p/demo/timelock"
+
+import (
+    "strconv"
+    "std"
+    "gno.land/p/demo/avl"
+    "gno.land/p/demo/accesscontrol"
+    "gno.land/p/demo/seqid"
+    "time"
+)
+
+// Marks operations completed
+const DONE_TIMESTAMP = 1
+
+// Represents the status of a planned operation
+type OperationState int
+
+const (
+    Unset OperationState = iota
+    Waiting
+    Ready
+    Done
+)
+
+// TimelockUtil stores the necessary parameters for the timelock operations
+type TimelockUtil struct {
+    timestamps    *avl.Tree
+    accessControl *accesscontrol.RoleData
+}
+
+// New instance of TimelockUtil
+func NewTimelockUtil(timestamps *avl.Tree, accessControl *accesscontrol.RoleData) *TimelockUtil {
+    return &TimelockUtil{
+        timestamps:    timestamps,
+        accessControl: accessControl,
+    }
+}
+
+// Schedules an operation to be carried out after a minimum delay
+func (tl *TimelockUtil) Schedule(id seqid.ID, delay int64, minDelay int64) {
+    if delay < minDelay {
+        panic("timelockutil: Schedule: insufficient delay")
+    }
+    if tl.timestamps.Has(id.Binary()) {
+        panic("timelockutil: Schedule: operation already scheduled")
+    }
+    tl.timestamps.Set(id.Binary(), time.Now().Unix()+delay)
+    std.Emit("CallScheduled", "id", id.String(), "delay", strconv.FormatInt(delay, 10))
+}
+
+// Cancels a planned operation
+func (tl *TimelockUtil) Cancel(id seqid.ID) {
+    if !tl.IsOperationPending(id) {
+        panic("timelock: Cancel: operation not pending")
+    }
+    tl.timestamps.Remove(id.Binary())
+    std.Emit("Cancelled", "id", id.String())
+}
+
+// Executes a ready operation
+func (tl *TimelockUtil) Execute(id seqid.ID) {
+    if !tl.IsOperationReady(id) {
+        panic("timelock: Execute: operation not ready")
+    }
+    tl.timestamps.Set(id.Binary(), DONE_TIMESTAMP)
+    std.Emit("CallExecuted", "id", id.String())
+}
+
+// Update the minimum lead time for future operations
+func (tl *TimelockUtil) UpdateDelay(minDelay *int64, newDelay int64) {
+    if std.GetOrigCaller() != tl.accessControl.GetRoleAdmin() {
+        panic("timelock: UpdateDelay: only admin can update delay")
+    }
+    std.Emit("MinDelayChange", "oldDelay", strconv.FormatInt(*minDelay, 10), "newDelay", strconv.FormatInt(newDelay, 10))
+    *minDelay = newDelay
+}
+
+// Checks if an operation is pending
+func (tl *TimelockUtil) IsOperationPending(id seqid.ID) bool {
+    state := tl.GetOperationState(id)
+    return state == Waiting || state == Ready
```

Comment:
What's the use case for including `Ready` here?  I'm thinking, if someone calls this method and it returns true, then they will have to call `IsOperationReady` to see if `IsOperationPending` returned true because it was `Waiting` or `Pending`. IMO it would be easier to make the methods simply `IsWaiting` and `IsPending`.

Reply:
This is modification [#2307](https://github.com/gnolang/gno/pull/2307/commits/35279e0cefa14362204997eba3d8cad9928d1f65)

We simply replace `Waiting` with `Pending`  because in the end it does the same thing.  And we delete `|| state == Ready` for evitate call the `IsOperationReady`.
Is this that you had in mind in your explanation ?